libraries/networking/src/AssetClient.h

```diff
@@ -33,18 +37,98 @@ struct AssetInfo {
     int64_t size;
 };
 
+using MappingOperationCallback = std::function<void(bool responseReceived, AssetServerError serverError, QSharedPointer<ReceivedMessage> message)>;
 using ReceivedAssetCallback = std::function<void(bool responseReceived, AssetServerError serverError, const QByteArray& data)>;
 using GetInfoCallback = std::function<void(bool responseReceived, AssetServerError serverError, AssetInfo info)>;
 using UploadResultCallback = std::function<void(bool responseReceived, AssetServerError serverError, const QString& hash)>;
 using ProgressCallback = std::function<void(qint64 totalReceived, qint64 total)>;
 
 
+class GetMappingRequest : public QObject {
+    Q_OBJECT
+public:
+    GetMappingRequest(AssetPath path);
+
+    Q_INVOKABLE void start();
+
+    AssetHash getHash() { return _hash;  }
+    AssetServerError getError() { return _error;  }
+
+signals:
+    void finished(GetMappingRequest* thisRequest);
+
+private:
+    AssetPath _path;
+    AssetHash _hash;
+    AssetServerError _error { AssetServerError::NoError };
+};
+
+class SetMappingRequest : public QObject {
+    Q_OBJECT
+public:
+    SetMappingRequest(AssetPath path, AssetHash hash);
+
+    Q_INVOKABLE void start();
+
+    AssetHash getHash() { return _hash;  }
+    AssetServerError getError() { return _error;  }
+
+signals:
+    void finished(SetMappingRequest* thisRequest);
+
+private:
+    AssetPath _path;
+    AssetHash _hash;
+    AssetServerError _error { AssetServerError::NoError };
+};
+
+class DeleteMappingRequest : public QObject {
+    Q_OBJECT
+public:
+    DeleteMappingRequest(AssetPath path);
+
+    Q_INVOKABLE void start();
+
+    AssetServerError getError() { return _error;  }
+
+signals:
+    void finished(DeleteMappingRequest* thisRequest);
+
+private:
+    AssetPath _path;
+    AssetServerError _error { AssetServerError::NoError };
+};
+
+class GetAllMappingsRequest : public QObject {
+    Q_OBJECT
+public:
+    GetAllMappingsRequest();
+
+    Q_INVOKABLE void start();
+
+    AssetServerError getError() { return _error;  }
+    AssetMapping getMappings() { return _mappings;  }
+
+signals:
+    void finished(GetAllMappingsRequest* thisRequest);
+
+private:
+    std::map<AssetPath, AssetHash> _mappings;
+    AssetServerError _error { AssetServerError::NoError };
+};
```

Comment:
You'd still need an implementation of start in the subclasses. You could pull out the thread checking into the superclass which would be nice. Get/Set/Delete don't all use hash and path so not really any help there, adds confusion.

I think it will be worth it to add a `MappingRequest` class so the errors can be namespaced under it, though. I'll update.